packages/token/src/tests/erc721/test_erc721_enumerable.cairo

```diff
@@ -0,0 +1,581 @@
+    contract_state.erc721.mint(OWNER(), TOKEN_3);
+
+    let expected_list = array![TOKEN_1, TOKEN_2, TOKEN_3];
+    assert_dual_token_by_index(expected_list.span());
+}
+
+//
+// token_of_owner_by_index
+//
+
+#[test]
+fn test_token_of_owner_by_index() {
+    let (_, tokens_list) = setup();
+
+    assert_dual_token_of_owner_by_index(OWNER(), tokens_list);
+}
+
+#[test]
+#[should_panic(expected: ('ERC721Enum: out of bounds index',))]
+fn test_token_of_owner_by_index_when_index_equals_owned_tokens() {
+    let (state, tokens_list) = setup();
+    let owned_token_len = tokens_list.len().into();
+
+    state.token_of_owner_by_index(OWNER(), owned_token_len);
+}
+
+#[test]
+#[should_panic(expected: ('ERC721Enum: out of bounds index',))]
+fn test_token_of_owner_by_index_when_index_exceeds_owned_tokens() {
+    let (state, tokens_list) = setup();
+    let owned_tokens_len_plus_one = tokens_list.len().into() + 1;
+
+    state.token_of_owner_by_index(OWNER(), owned_tokens_len_plus_one);
+}
+
+#[test]
+#[should_panic(expected: ('ERC721Enum: out of bounds index',))]
+fn test_token_of_owner_by_index_when_target_has_no_tokens() {
+    let (state, _) = setup();
+
+    state.token_of_owner_by_index(OTHER(), 0);
+}
+
+#[test]
+#[should_panic(expected: ('ERC721: invalid account',))]
+fn test_token_of_owner_by_index_when_owner_is_zero() {
+    let (state, _) = setup();
+
+    state.token_of_owner_by_index(ZERO(), 0);
+}
+
+#[test]
+fn test_token_of_owner_by_index_remove_last_token() {
+    let (_, tokens_list) = setup();
+    let mut contract_state = CONTRACT_STATE();
+    let last_token = *tokens_list.at(tokens_list.len() - 1);
+
+    contract_state.erc721.transfer(OWNER(), RECIPIENT(), last_token);
+
+    let expected_list = array![TOKEN_1, TOKEN_2];
+    assert_dual_token_of_owner_by_index(OWNER(), expected_list.span());
+}
+
+#[test]
+fn test_token_of_owner_by_index_remove_first_token() {
+    let (_, tokens_list) = setup();
+    let mut contract_state = CONTRACT_STATE();
+    let first_token = *tokens_list.at(0);
+
+    contract_state.erc721.transfer(OWNER(), RECIPIENT(), first_token);
+
+    // Removed tokens are replaced by the last token
+    // to prevent indexing gaps
+    let expected_list = array![TOKEN_3, TOKEN_2];
+    assert_dual_token_of_owner_by_index(OWNER(), expected_list.span());
+}
+
+#[test]
+fn test_token_of_owner_by_index_when_all_tokens_transferred() {
+    let (_, tokens_list) = setup();
+    let mut contract_state = CONTRACT_STATE();
+
+    contract_state.erc721.transfer(OWNER(), RECIPIENT(), TOKEN_1);
+    contract_state.erc721.transfer(OWNER(), RECIPIENT(), TOKEN_2);
+    contract_state.erc721.transfer(OWNER(), RECIPIENT(), TOKEN_3);
+
+    assert_dual_token_of_owner_by_index(RECIPIENT(), tokens_list);
+}
+
+//
+// _update
+//
+
+#[test]
+fn test__update_when_mint() {
+    let (mut state, _) = setup();
+    let initial_supply = state.total_supply();
+    let new_token = 'TOKEN_4';
+
+    state.before_update(OWNER(), new_token);
```

Comment:
Should we call mint/burn in the tests instead of calling `before_update` directly? These calls lead to inconsistent state of the component: enumerable mappings get updated, but the balance of the user does not

Reply:
These tests are specifically testing the `before_update` fn. The rest of the test suite uses mint and burn. I'll update the `before_update` test names to make this more clear